Add SortResponseProcessor to Search Pipelines

dbwiddis opened the pull request:
### Description

Adds a new `SearchResponseProcessor` implementation that sorts a field, comparable to the Ingest Pipeline [Sort Processor](https://opensearch.org/docs/latest/ingest-pipelines/processors/sort/#:~:text=The%20sort%20processor%20sorts%20an,input%20is%20not%20an%20array.).

### Related Issues

Part 1 (of 2) of #14758

### Documentation PR (for both this and #14800)

- https://github.com/opensearch-project/documentation-website/pull/7767
  - JSON examples were the result of actual testing of this PR code
- https://github.com/opensearch-project/opensearch-api-specification/pull/440

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dbwiddis commented:
Flaky test is tracked here:
 - https://github.com/opensearch-project/OpenSearch/issues/14318

owaiskazi19 commented on modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/SortResponseProcessor.java:
```diff
@@ -0,0 +1,212 @@
+        return response;
+    }
+
+    private <T extends Comparable<T>> List<Object> getSortedValues(List<Object> values) {
+        List<T> comparableValues = new ArrayList<>(values.size());
+        for (Object obj : values) {
+            if (obj == null) {
+                throw new IllegalArgumentException("field [" + sortField + "] contains a null value.]");
+            } else if (Comparable.class.isAssignableFrom(obj.getClass())) {
+                @SuppressWarnings("unchecked")
+                T comp = (T) obj;
+                comparableValues.add(comp);
+            } else {
+                throw new IllegalArgumentException(
+                    "field [" + sortField + "] of type [" + obj.getClass().getName() + "] is not comparable.]"
+                );
+            }
+        }
+        return comparableValues.stream()
+            .sorted(sortOrder.equals(SortOrder.ASCENDING) ? Comparator.naturalOrder() : Comparator.reverseOrder())
+            .collect(Collectors.toList());
+    }
```
We can split the methods into two here. Reasons:
1. More readability
2. Can use Collections.sort which in turns sort the list in-place and same like Ingest pipeline's split processor
```suggestion
    private List<? extends Comparable<Object>> getComparableValues(List<Object> values) {
    List<Comparable<Object>> comparableValues = new ArrayList<>(values.size());
    for (Object obj : values) {
        if (obj == null) {
            throw new IllegalArgumentException("field [" + sortField + "] contains a null value.");
        } else if (obj instanceof Comparable) {
            @SuppressWarnings("unchecked")
            Comparable<Object> comp = (Comparable<Object>) obj;
            comparableValues.add(comp);
        } else {
            throw new IllegalArgumentException(
                "field [" + sortField + "] of type [" + obj.getClass().getName() + "] is not comparable."
            );
        }
    }
    return comparableValues;
}

private List<? extends Comparable<Object>> sortValues(List<? extends Comparable<Object>> values) {
    List<Comparable<Object>> copy = new ArrayList<>(values);
    if (sortOrder.equals(SortOrder.ASCENDING)) {
        Collections.sort(copy);
    } else {
        Collections.sort(copy, Collections.reverseOrder());
    }
    return copy;
}
```

dbwiddis replied:
I actually did have it as two separate methods, but since the second method can be written as a one-liner with streams I combined them.

I agree that including a ternary operator in the middle of a stream strains readability, though, so I'll go back to the two-phase method.

dbwiddis replied:
Actually, it's not that simple.  While you can sort in-place, the return type necessary is `List<Object>` to be used in the call to the `DocumentField` constructor.  And you can't change the type of an object in-place.

I'll still separate the methods for readability, but there will still be a stream at the end to do the typecast.

owaiskazi19 replied:
Can you access them something like the below?
```
List<? extends Comparable<Object>> values = getComparableValues(docField.getValues());
List<? extends Comparable<Object>> sortedValues = sortValues(values);
```

dbwiddis replied:
> Can you access them something like the below?

`List<? extends Comparable<Object>> sortedValues = ...` does not compile with `new DocumentField(targetField, sortedValues)`.

owaiskazi19 commented on modules/search-pipeline-common/src/test/java/org/opensearch/search/pipeline/common/SortResponseProcessorTests.java:
```diff
@@ -0,0 +1,230 @@
+            () -> factory.create(Collections.emptyMap(), null, null, false, Collections.emptyMap(), null)
+        );
+    }
+}
```
Do we need more tests for different data types sort as well? For eg: SortFloat, SortDoubles etc?

dbwiddis replied:
any class that implements `Comparable` (which all the `Number` classes do, as well as `String`) will work with the existing code, so the underlying class type doesn't matter.

owaiskazi19 reviewed (commented):
Overall LG. Added few comments

jackiehanyang commented:
LGTM, thanks for the change!

andrross commented on modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/SortResponseProcessor.java:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    private List<? extends Comparable<Object>> getComparableValues(List<Object> values) {
+        List<Comparable<Object>> comparableValues = new ArrayList<>(values.size());
```
Do you actually have to make a copy here? It looks like this list is being fully copied twice. Once seems necessary as the sort must mutate the list, but it seems like it should be possible to avoid the second copy.

andrross replied:
(You do in fact need to copy the list before the sort, so it might be the copy on line 177 that can be avoided)

dbwiddis replied:
> Do you actually have to make a copy here?

@andrross I'm pretty sure we do, although I'd be happy to be shown I'm wrong.

The underlying issue is typecasting. 

We start out with a `List<Object>` coming from [`DocumentField.getValues()`](https://github.com/opensearch-project/OpenSearch/blob/9c6e6187864c976656889cdec45064920cf856ee/server/src/main/java/org/opensearch/common/document/DocumentField.java#L98-L100).   There are a finite number of types possible here, and not all of them are sortable (specifically nested maps or lists).

So I do need to iterate the list of `Object` and type-check that they implement `Comparable` before casting and reassembling the list and sorting it.

However, the [DocumentField constructor](https://github.com/opensearch-project/OpenSearch/blob/9c6e6187864c976656889cdec45064920cf856ee/server/src/main/java/org/opensearch/common/document/DocumentField.java#L73-L76) we use to replace (or add new) the value also expects a `List<Object>`.

[We can't just pass the subclass list and expect Java to handle it](https://stackoverflow.com/a/2745301/1161484).

In my original PR submission, I combined the sort and this second "copy" by using a stream, as   `collect(Collectors.toList())` is smart enough to do the needed typecasting.   However, @owaiskazi19 thought it was a lot less readable so I split it up.   See https://github.com/opensearch-project/OpenSearch/pull/14785#discussion_r1681566351  But it still requires some manipulation of the list to compile when a `List<Object>` is expected.

However, I see that [`List.copyOf(comparableList)`](https://stackoverflow.com/a/72196150/1161484) might work as well; ironically a `copyOf()` when we want to avoid a copy but I think it does "copy in place" in this particular case.

dbwiddis replied:
> However, I see that [`List.copyOf(comparableList)`](https://stackoverflow.com/a/72196150/1161484) might work as well; ironically a `copyOf()` when we want to avoid a copy but I think it does "copy in place" in this particular case.

Actually, [reading further](https://stackoverflow.com/questions/5082044/most-efficient-way-to-cast-listsubclass-to-listbaseclass/72195980#72195980) a copy of the array is made. [JDK Source here](https://github.com/openjdk/jdk/blob/c25c4896ad9ef031e3cddec493aef66ff87c48a7/src/java.base/share/classes/java/util/ImmutableCollections.java#L168-L176). The only time it isn't is if the original list was created with `List.of()` or if it's empty.

Ultimately it then boils down to syntax. `List.copyOf()` may look more elegant than a new ArrayList, but since we don't care about immutability I'm not sure it matters.

I liked my original stream implementation better. ;)

andrross replied:
I think the alternative is to do an unchecked cast `return (List)comparableValues;` on line 177 to avoid the extra copy on the way back. Is that safe?

dbwiddis replied:
> I think the alternative is to do an unchecked cast `return (List)comparableValues;` on line 177 to avoid the extra copy on the way back. Is that safe?

From the narrow perspective of a `SearchResponseProcessor`, the answer is "we don't know, and shouldn't make assumptions about anything outside our code."

The JDK source linked above indicates the "safety" decision that `List.copyOf()` uses to decide whether to cast without copying:
>  unless the arg is already a safe, null-prohibiting unmodifiable list

It's not that type of list, but we'd need to know how it was going to be used by the Search Pipeline.  

The suggestion does bypass type safety; the return value would really be a `List<? extends Comparable<Object>>` at runtime.  Since we don't really control the code consuming the return value, I'd argue that we're relying on implementation details of a different class just to save us the overhead of copying an array.

I'm pretty sure the only "unsafe" thing that could happen would be if someone tried to add something to the list later assuming they could add any `Object`, but actually added something that was not comparable or was `null`.   We happen to know that's not the case for Search Pipelines but only because we can read the code.

We could maybe wrap this unchecked cast in a read-only wrapper to prevent the copy and prevent inserting an incorrect object later, but we're really getting into micro-optimization here, and I'm really not sure it's worth it.

I think my current `List.copyOf(comparableValue)` is the most correct, and before I'd violate type safety I'd want a really good performance-based reason with benchmarks showing it's worth it.  My brief research into the performance of `Arrays.copyOf()` indicates two orders of magnitude smaller time to copy a list than for each comparison in the sort of that list.

dbwiddis replied:
> Do you actually have to make a copy here? It looks like this list is being fully copied twice. Once seems necessary

@andrross @owaiskazi19 You both have made good comments regarding creating new objects.  I've been thinking about this and realized that while one list copy is necessary (for the downcast), it was the the logical separation of downcasting a whole list before sorting that created the need to do something less than optimal after sorting the list (copy or unchecked cast).   However, by simply refactoring to do the downcast by element (which we did anyway with the list method inside a loop) we can simplify it to a single new list creation, keep the parts of the process logically separated for readability, and keep the elegance of the streams that I probably like more than I should. ;)

Here's the result in the latest commit. I hope it satisfies both readability and new-object-creation concerns.  LMK if you need any further changes.

```java
    private List<Object> getSortedValues(List<Object> values) {
        return values.stream()
            .map(this::downcastToComparable)
            .sorted(sortOrder.equals(SortOrder.ASCENDING) ? Comparator.naturalOrder() : Comparator.reverseOrder())
            .collect(Collectors.toList());
    }

    @SuppressWarnings("unchecked")
    private Comparable<Object> downcastToComparable(Object obj) {
        if (obj == null) {
            throw new IllegalArgumentException("field [" + sortField + "] contains a null value.]");
        } else if (Comparable.class.isAssignableFrom(obj.getClass())) {
            return (Comparable<Object>) obj;
        } else {
            throw new IllegalArgumentException("field [" + sortField + "] of type [" + obj.getClass().getName() + "] is not comparable.]");
        }
    }
```

owaiskazi19 reviewed (approved):
LGTM. Thanks for addressing the comments patiently

dbwiddis commented:
> LGTM. Thanks for addressing the comments patiently

Thank you for educating me!  What I thought were simple PRs required some creativity to minimize the copying while properly casting.

Fun fact, did you know `docvalue_fields` stores lists in a pre-sorted order for performance?  That was fun hunting down that (non-)bug.

dbwiddis commented:
Flaky test tracked by:
 - https://github.com/opensearch-project/OpenSearch/issues/14551

dbwiddis commented:
Flaky (unstable) test tracked by:
 - https://github.com/opensearch-project/OpenSearch/issues/9401

Merging!

dbwiddis commented:
Should have waited to merge until backport of the Split processor was merged.... I will wait to manually backport until then.
